lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+            # that can work on the current package.
+            try:
+                self._install_task(task, **kwargs)
+                self._update_installed(task)
+
+                # If we installed then we should keep the prefix
+                last_phase = getattr(pkg, 'last_phase', None)
+                keep_prefix = last_phase is None or keep_prefix
+
+            except spack.directory_layout.InstallDirectoryAlreadyExistsError:
+                tty.warn("Keeping existing install prefix in place.")
+
+                # TODO: Does "best effort" mean this is considered installed?
+                self._update_installed(task)
+
+                # TODO: Does "best effort" mean raise exception here?
+                raise
+
+            except (Exception, KeyboardInterrupt, SystemExit) as exc:
+                # Assuming best effort installs so suppress the exception and
+                # mark as a failure UNLESS this is the explicit package.
+                # TODO: Does this interplay properly with Spack's
+                # TODO: 'fail_on_exit' option given the above goal?
+                err = 'Failed to install {0} due to {1}: {2}'
+                tty.error(err.format(pkg.name, exc.__class__.__name__,
+                          str(exc)))
+                self._update_failed(task, True, exc)
+
+                if pkg_id == self.pkg.unique_id:
+                    # TODO: Prefer the following but creates challenges for
+                    # TODO: updating some existing tests.
+                    # reason = 'due to {0}: {1}'.format(exc.__class__.__name__,
+                    #                                   str(exc))
+                    # raise InstallError('Installation of {0} failed {1}'
+                    #                    .format(pkg_id, reason))
+                    raise
+
+            finally:
+                # Remove the install prefix if anything went wrong during
+                # install.
+                if not keep_prefix:
+                    pkg.remove_prefix()
+
+                # The subprocess *may* have removed the build stage. Mark it
+                # not created so that the next time pkg.stage is invoked, we
+                # check the filesystem for it.
+                pkg.stage.created = False
+
+            # Perform basic task cleanup for the installed spec to
+            # include downgrading the write to a read lock
+            self._cleanup_task(pkg, True)
+
+        # Cleanup, which includes releasing all of the read locks
+        self._cleanup_all_tasks()
+
+        # Ensure we properly report if the original/explicit pkg is failed
+        if self.pkg.unique_id in self.failed:
+            msg = ('Installation of {0} failed.  Review log for details'
+                   .format(self.pkg.unique_id))
+            tty.error(msg)
+            raise InstallError(msg)
+
+    install.__doc__ += install_args_docstring
```

Comment:
Eh, not sure how much this will help somebody looking for docs or parses in IDE that show docs automatically...

Reply:
This was the approach Peter (scheibelp) used in previous commits to ensure the documentation included the arguments in both locations.  I assumed this was the desired approach to use in this case.